zachsinterview/src/main/java/com/imprev/zachsinterview/NumericalKeyValueStoreImpl.java

```
@@ -0,0 +1,111 @@
+        try {
+            this.thirdPartyAPIClient.verifyCredentials();
+        } catch (IOException | TryAgainLaterException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void put(String key, Integer value) throws NumericalKeyValueException {
+
+        int retries = 0;
+        String errorMessage = null;
+        while (retries < MAX_RETRIES) {
+            try {
+                // we will only store one value in the list data structure the third party API uses, so if a
+                // value already exists we want to override it not append to the list.
+                List<String> values = this.thirdPartyAPIClient.get(key);
+                
+                if (values.size() == 1) {
+                    this.thirdPartyAPIClient.delete(key);
+                }
+                
+                this.thirdPartyAPIClient.add(key, String.valueOf(value));
+                
+                return;
+
+            } catch (IOException | TryAgainLaterException | InvalidCredentialsException e) {
+                errorMessage = e.getMessage();
+                retries++;
+            }
+        }
+        throw new NumericalKeyValueException(errorMessage);
+    }
+
+    @Override
+    public void delete(String key) throws NumericalKeyValueException {
+        try {
+            this.thirdPartyAPIClient.delete(key);
+            
+        } catch (IOException | TryAgainLaterException | InvalidCredentialsException e) {
+            throw new NumericalKeyValueException(e.getMessage());
+        }
+    }
+
+    @Override
+    public Integer get(String key) throws NumericalKeyValueException {
+        int retries = 0;
+        String errorMessage = null;
+        while (retries < MAX_RETRIES) {
+            try {
+                // we will only store one value in the list data structure the third party API uses
+                return Integer.parseInt(this.thirdPartyAPIClient.get(key).get(0));
```

Need to handle if Key doesn't exist. Check if key exists in the map and return null if it doesn't
We should add a containsKey method to the interface to allow the users to verify if a Key exists before attempting to retrieve it. Once that is added, we should return an exception on gets where the key doesn't exist.